lib/mayaUsd/ufe/StagesSubject.cpp

```diff
@@ -120,10 +174,71 @@ void valueChanged(const Ufe::Path& ufePath, const TfToken& changedToken)
             pendingAttributeChangedNotifications.emplace_back(p);
         }
     } else {
-        sendValueChanged(ufePath, changedToken);
+        sendAttributeChanged(ufePath, changedToken, AttributeChangeType::valueChanged);
     }
 }
 
+void attributeChanged(
+    const Ufe::Path&    ufePath,
+    const TfToken&      changedToken,
+    AttributeChangeType changeType)
+{
+    if (inAttributeChangedNotificationGuard()) {
+        // Don't add pending notif if one already exists with same path/token.
+        auto p = AttributeNotification { ufePath, changedToken, changeType };
+        if (std::find(
+                pendingAttributeChangedNotifications.begin(),
+                pendingAttributeChangedNotifications.end(),
+                p)
+            == pendingAttributeChangedNotifications.end()) {
+            pendingAttributeChangedNotifications.emplace_back(p);
+        }
+    } else {
+        sendAttributeChanged(ufePath, changedToken, changeType);
+    }
+}
+
+void processAttributeChanges(
+    const Ufe::Path&                                ufePath,
+    const SdfPath&                                  changedPath,
+    const std::vector<const SdfChangeList::Entry*>& UFE_V4_24(entries))
+{
+#if (UFE_PREVIEW_VERSION_NUM >= 4024)
+    bool sendValueChanged = false;
+    bool sendAdded = false;
+    bool sendRemoved = false;
+    bool sendConnectionChanged = false;
+    for (const auto& entry : entries) {
+        if (entry->flags.didAddProperty || entry->flags.didAddPropertyWithOnlyRequiredFields) {
+            sendAdded = true;
+        } else if (
+            entry->flags.didRemoveProperty
+            || entry->flags.didRemovePropertyWithOnlyRequiredFields) {
+            sendRemoved = true;
+        } else if (entry->flags.didChangeAttributeConnection) {
+            sendConnectionChanged = true;
+        } else {
+            sendValueChanged = true;
+        }
+    }
```

Comment:
Here you have a loop so its possible we will send more than one notification now, such as attributeChanged and valueChanged, whereas the old code would send a single valueChanged. Is this correct?

Reply:
Yes, but with finer grained notifications we need at least one of each type as this loop currently does. And even there I am not sure we should collapse topological changes. A sequence like add-remove-add attribute should optimally result in a single AttributeAdded sent.